src/sql_utils.h

```diff
@@ -0,0 +1,123 @@
+#ifndef SQL_UTILS_H_
+#define SQL_UTILS_H_
+
+#include <memory>
+
+#include <sqlite3.h>
+
+// Unique ownership SQLite3 statement creation
+template <typename T>
+static void bindArguments(sqlite3* db, sqlite3_stmt* statement, int cnt, int v) {
+  if (sqlite3_bind_int(statement, cnt, v) != SQLITE_OK) {
+    LOG_ERROR << "Could not bind: " << sqlite3_errmsg(db);
+    throw std::runtime_error("SQLite bind error");
+  }
+}
+
+template <typename T>
+static void bindArgument(sqlite3* db, sqlite3_stmt* statement, int cnt, const char* v) {
+  if (sqlite3_bind_text(statement, cnt, v, -1, nullptr) != SQLITE_OK) {
+    LOG_ERROR << "Could not bind: " << sqlite3_errmsg(db);
+    throw std::runtime_error("SQLite bind error");
+  }
+}
+
+template <typename T>
+static void bindArgument(sqlite3* db, sqlite3_stmt* statement, int cnt, const std::string& v) {
+  bindArgument<const char*>(db, statement, cnt, v.c_str());
+}
+
+struct SQLBlob {
+  const std::string& content;
+  SQLBlob(const std::string& str) : content(str) {}
+};
+
+template <typename T>
+static void bindArgument(sqlite3* db, sqlite3_stmt* statement, int cnt, const SQLBlob& blob) {
+  if (sqlite3_bind_blob(statement, cnt, blob.content.c_str(), blob.content.size(), NULL) != SQLITE_OK) {
+    LOG_ERROR << "Could not bind: " << sqlite3_errmsg(db);
+    throw std::runtime_error("SQLite bind error");
+  }
+}
+
+struct SQLZeroBlob {
+  size_t size;
+};
+
+template <typename T>
+static void bindArgument(sqlite3* db, sqlite3_stmt* statement, int cnt, const SQLZeroBlob& blob) {
+  if (sqlite3_bind_zeroblob(statement, cnt, blob.size) != SQLITE_OK) {
+    LOG_ERROR << "Could not bind: " << sqlite3_errmsg(db);
+    throw std::runtime_error("SQLite bind error");
+  }
+}
+
+template <typename... Types>
+static void bindArguments(sqlite3* db, sqlite3_stmt* statement, int cnt) {
+  /* end of template specialization */
+  (void)db;
+  (void)statement;
+  (void)cnt;
+}
+
+template <typename T, typename... Types>
+static void bindArguments(sqlite3* db, sqlite3_stmt* statement, int cnt, const T& v, const Types&... args) {
+  bindArgument<const T&>(db, statement, cnt, v);
+  bindArguments<const Types&...>(db, statement, cnt + 1, args...);
+}
+
+class SQLiteStatement {
```

Comment:
Should this be boost::noncopyable?

Reply:
It is actually enforced automatically because the class has a non-copyable member (unique_ptr...).

But it might be nice to use it anyway as some kind of annotation.

Reply:
No, it should be constructor = delete since we have C++11! No need for boost.

Reply:
So, should we mandate that the copy constructor and assignment operator be deleted explicitly even if that's already done by the compiler?
(see: https://google.github.io/styleguide/cppguide.html#Copyable_Movable_Types)

Reply:
I don't have a firm opinion here. If you know the compiler is already doing the right thing, then it probably doesn't matter except for making it obvious to anyone reading the code.